shopfloor/services/cluster_picking.py

```diff
@@ -1122,17 +1122,19 @@ def set_destination_all(self, picking_batch_id, barcode, confirmation=False):
     def _unload_write_destination_on_lines(self, lines, location):
         lines.write({"shopfloor_unloaded": True, "location_dest_id": location.id})
         lines.package_level_id.location_dest_id = location
-        for line in lines:
-            # We set the picking to done only when the last line is
-            # unloaded to avoid backorders.
-            picking = line.picking_id
-            if picking.state == "done":
-                continue
-            picking_lines = picking.mapped("move_line_ids")
-            if all(line.shopfloor_unloaded for line in picking_lines):
-                picking._action_done()
+        for picking in lines.batch_id.picking_ids:
+            picking_lines = lines.filtered(lambda l, p=picking: l.picking_id == p)
+            self._unload_set_picking_to_done(picking, picking_lines)
+
+    def _unload_set_picking_to_done(self, picking, picking_lines):
+        if picking.state == "done":
+            return
+        # We set the picking to done only when the last line is
+        # unloaded to avoid backorders.
+        if all(line.shopfloor_unloaded for line in picking.move_line_ids):
+            picking._action_done()
         if self.work.menu.unload_package_at_destination:
-            lines.result_package_id = False
+            picking_lines.result_package_id = False
```

Comment:
@sbejaoui It's difficult to know what's the semantic behind this new method '_unload_set_picking_to_done'. In your change, you still iterate multi times n the lines.....



```python
from collections import namedtuple

    def _unload_write_destination_on_lines(self, lines, location):
        lines.write({"shopfloor_unloaded": True, "location_dest_id": location.id})
        lines.package_level_id.location_dest_id = location
        LineIdsByPickingValue =  namedtuple("LINESBYP", ["ids", "all_unloaded_status"])
       bypicking = {} 
       for line in lines:
           value = lines_by_picking_value[line.picking_id]
           if value:
               value = LineIdsByPickingValue(ids = value.ids + line.ids, all_unloaded_status = value.all_unloaded_status and line.shopfloor_unloaded)
           else:
               value = LineIdsByPickingValue(ids=line.ids, all_unloaded_status=line.shopfloor_unloaded)
           lines_by_picking_value[line.picking_id] = value

    for picking, line_values in lines_by_picking_value.items():
        if line_values.all_unloaded_status:
            picking._action_done()
       if self.work.menu.unload_package_at_destination:
           self.env["stock.move.line"].browse(line_values.ids).write({"result_package_id" : False})
     
```

```